File: scripts/flutter/keystore/android_show_info_keystore.py
#!/usr/bin/env python3
"""Keystore 정보 조회 스크립트."""

import argparse
import subprocess
import sys
from pathlib import Path

def parse_args() -> argparse.Namespace:
    """CLI 인자를 파싱합니다."""
    parser = argparse.ArgumentParser(description="Keystore 정보 조회")
    parser.add_argument(
        "keystore",
        help="키스토어 파일 경로 (예: my-release-key.keystore)",
    )
    parser.add_argument(
        "--project-dir",
        required=True,
        help="프로젝트 루트 디렉토리",
    )
    return parser.parse_args()


def find_android_dir(project_root: Path) -> Path | None:
    """프로젝트 루트에서 android 폴더를 찾습니다."""
    # 직접 android 폴더가 있는 경우
    if (project_root / "android").is_dir():
        return project_root / "android"

    # 하위 디렉토리에서 android 폴더 찾기 (1단계만)
    for child in project_root.iterdir():
        if child.is_dir() and (child / "android").is_dir():
            return child / "android"

    return None


def main() -> int:
    """키스토어 정보를 출력합니다."""
    args = parse_args()

    # 키스토어 경로 결정
    keystore_path = Path(args.keystore)
    if not keystore_path.is_absolute():
        # 상대 경로인 경우 프로젝트 루트 기준으로 android 폴더를 찾아서 설정
        project_root = Path(args.project_dir)
        android_dir = find_android_dir(project_root)

        if android_dir is None:
            print("Error: android directory not found in project", file=sys.stderr)
            return 1

        keystore_path = android_dir / args.keystore

    if not keystore_path.exists():
        print(f"Error: Keystore not found at {keystore_path}", file=sys.stderr)
        return 1

    cmd = [
        "keytool",
        "-list",
        "-keystore",
        str(keystore_path),
        "-v",
    ]

    try:
        result = subprocess.run(cmd, check=True)
        return result.returncode
    except subprocess.CalledProcessError as e:
        return e.returncode
    except FileNotFoundError:
        print(
            "Error: keytool command not found. Please ensure JDK is installed.",
            file=sys.stderr,
        )
        return 1


if __name__ == "__main__":
    sys.exit(main())

File: scripts/flutter/keystore/android-show-info-keystore.mjs
#!/usr/bin/env node
// =============================================================================
// Keystore 정보 조회 스크립트.
//
// Usage:
//   android-show-info-keystore.mjs <keystore> --project-dir <dir>
// =============================================================================

import fs from 'node:fs';
import path from 'node:path';
import process from 'node:process';
import { spawnSync } from 'node:child_process';

const HELP = `Usage: android-show-info-keystore.mjs <keystore> --project-dir <dir>

Keystore 정보를 조회합니다 (keytool -list -v).

Arguments:
  <keystore>           키스토어 파일 경로 (예: my-release-key.keystore)

Options:
  --project-dir <dir>  프로젝트 루트 디렉토리 (required)
  -h, --help           Show this help
`;

function usage(code = 1) {
  (code === 0 ? process.stdout : process.stderr).write(HELP);
  process.exit(code);
}

function parseArgs(argv) {
  const args = { keystore: '', projectDir: '' };
  const positional = [];
  const rest = argv.slice(2);

  while (rest.length > 0) {
    const a = rest.shift();
    switch (a) {
      case '--project-dir':
        if (!rest.length) {
          process.stderr.write('--project-dir requires a value\n');
          usage();
        }
        args.projectDir = rest.shift();
        break;
      case '-h':
      case '--help':
        usage(0);
        break;
      default:
        if (a.startsWith('-')) {
          process.stderr.write(`Unknown option: ${a}\n`);
          usage();
        }
        positional.push(a);
    }
  }

  if (positional.length < 1) {
    process.stderr.write('Error: <keystore> is required\n');
    usage();
  }
  args.keystore = positional[0];

  if (!args.projectDir) {
    process.stderr.write('Error: --project-dir is required\n');
    usage();
  }

  return args;
}

function findAndroidDir(projectRoot) {
  const directAndroid = path.join(projectRoot, 'android');
  if (fs.existsSync(directAndroid) && fs.statSync(directAndroid).isDirectory()) {
    return directAndroid;
  }

  let entries;
  try {
    entries = fs.readdirSync(projectRoot, { withFileTypes: true });
  } catch {
    return null;
  }

  for (const entry of entries) {
    if (!entry.isDirectory()) continue;
    const candidate = path.join(projectRoot, entry.name, 'android');
    if (fs.existsSync(candidate) && fs.statSync(candidate).isDirectory()) {
      return candidate;
    }
  }

  return null;
}

function main() {
  const args = parseArgs(process.argv);

  let keystorePath = args.keystore;
  if (!path.isAbsolute(keystorePath)) {
    const projectRoot = path.resolve(args.projectDir);
    const androidDir = findAndroidDir(projectRoot);
    if (androidDir === null) {
      process.stderr.write('Error: android directory not found in project\n');
      process.exit(1);
    }
    keystorePath = path.join(androidDir, args.keystore);
  }

  if (!fs.existsSync(keystorePath)) {
    process.stderr.write(`Error: Keystore not found at ${keystorePath}\n`);
    process.exit(1);
  }

  const cmd = ['keytool', '-list', '-keystore', keystorePath, '-v'];
  const result = spawnSync(cmd[0], cmd.slice(1), { stdio: 'inherit' });

  if (result.error) {
    if (result.error.code === 'ENOENT') {
      process.stderr.write(
        'Error: keytool command not found. Please ensure JDK is installed.\n',
      );
      process.exit(1);
    }
    process.stderr.write(`Error: ${result.error.message}\n`);
    process.exit(1);
  }

  process.exit(result.status ?? 1);
}

main();
